osu.Game/OsuGame.cs

```diff
@@ -651,10 +651,14 @@ public void PerformFromScreen(Action<IScreen> action, IEnumerable<Type> validScr
             Add(performFromMainMenuTask = new PerformFromMenuRunner(action, validScreens, () => ScreenStack.CurrentScreen));
         }
 
-        public override void AttemptExit()
+        public override void AttemptExit(Action invokeWhenExiting = null)
         {
             // Using PerformFromScreen gives the user a chance to interrupt the exit process if needed.
-            PerformFromScreen(menu => menu.Exit());
+            PerformFromScreen(menu =>
+            {
+                invokeWhenExiting?.Invoke();
+                menu.Exit();
```

Comment:
Tested with `osu-deploy` by hardcoding version to an older date and making a custom build that updates to the latest release.

This doesn't seem to work. `menu.Exit()` calls `MainMenu.OnExiting()` / opens the confirmation dialog:
https://github.com/ppy/osu/blob/9724c6bb5350bb4e19731d4baadf6c96ca9c1514/osu.Game/Screens/Menu/MainMenu.cs#L280-L299

which means it'll still set `RestartWhenAppExited` before the dialog is even opened.

Reply:
Right.. that's one level higher than I expected. Bit of a pain to get this right. Maybe the menu should just set the flag itself somehow..